packages/transactions/src/decompile-transaction.ts

```diff
@@ -0,0 +1,181 @@
+import { assertIsAddress, Base58EncodedAddress } from '@solana/addresses';
+import { pipe } from '@solana/functional';
+import { AccountRole, IAccountMeta, IInstruction } from '@solana/instructions';
+import { Ed25519Signature } from '@solana/keys';
+
+import { Blockhash, setTransactionLifetimeUsingBlockhash } from './blockhash';
+import { getCompiledTransaction } from './compile-transaction';
+import { createTransaction } from './create-transaction';
+import { isAdvanceNonceAccountInstruction, Nonce, setTransactionLifetimeUsingDurableNonce } from './durable-nonce';
+import { setTransactionFeePayer } from './fee-payer';
+import { appendTransactionInstruction } from './instructions';
+import { CompiledMessage } from './message';
+import { ITransactionWithSignatures } from './signatures';
+import { TransactionVersion } from './types';
+
+type CompiledTransaction = ReturnType<typeof getCompiledTransaction>;
+type CompilableTransaction = Parameters<typeof getCompiledTransaction>[0];
+
+function getAccountMetas(message: CompiledMessage): IAccountMeta[] {
+    const { header } = message;
+    const numWritableSignerAccounts = header.numSignerAccounts - header.numReadonlySignerAccounts;
+    const numWritableNonSignerAccounts =
+        message.staticAccounts.length - header.numSignerAccounts - header.numReadonlyNonSignerAccounts;
+
+    const accountMetas: IAccountMeta[] = [];
+
+    let accountIndex = 0;
+    for (let i = 0; i < numWritableSignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.WRITABLE_SIGNER,
+        });
+        accountIndex++;
+    }
+
+    for (let i = 0; i < header.numReadonlySignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.READONLY_SIGNER,
+        });
+        accountIndex++;
+    }
+
+    for (let i = 0; i < numWritableNonSignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.WRITABLE,
+        });
+        accountIndex++;
+    }
+
+    for (let i = 0; i < header.numReadonlyNonSignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.READONLY,
+        });
+        accountIndex++;
+    }
+
+    return accountMetas;
+}
+
+function convertInstruction(
+    instruction: CompiledMessage['instructions'][0],
```

Comment:
This actually works for a type definition? Haha, I figured you'd need to work `typeof` in there somehow for an array.

Reply:
Yep! I'm guessing it's a result of the same weirdness that typescript never types an array index as undefined. Annoying there, handy here lol